### ISSUE TITLE
Is there a way to map a specific literal sequence of keystrokes to a shortcut key for use in command line

### ISSUE THREAD

**Issue by TryerGit** [VSCodeVim/Vim#5917]
I would like to run the command over a visual selection thus:

`:'<,'>sort /\k\S*(/ ri`

which sorts a visual selection based on the word just before a `(`

Is it possible to map the _literal keystrokes_ `sort /\k\S*(/ ri` to a key, say `Ctrl-Shift-s` so that the user can use this in different contexts instead of typing this out each time?

For instance, I would want to do the following keystrokes to obtain the above sort:

(KeyStrokeSet1) Visually select range of lines
(KeyStrokeSet2) `:`
(KeyStrokeSet3) `Ctrl-Shift-s`
[[At this point, the command line should display `:'<,'>sort /\k\S*(/ ri`]]
(KeyStrokeSet4) `Enter` to execute the sort over the visual selection.

In pure vim, this could be possible using `cnoremap` in the `.vimrc` (see comment [here](https://stackoverflow.com/questions/65404127/sort-class-member-functions-and-variables-in-ascending-order-in-header-file)) 

What is the equivalent way in VSCodeVim ?

Thanks.

**Comment by berknam** [comment]
Currently you can't do that with `<C-S-s>` because the extension is not capturing that key combination. You can do it with other key combinations though like `<C-k>` for example, like this:
```jsonc
    "vim.commandLineModeKeyBindingsNonRecursive": [
        {
            "before": ["<C-k>"],
            "after": ["s", "o", "r", "t", " ", "/", "\\", "k", "\\", "S", "*", "(", "/", " ", "r", "i"],
        },
    ],
```
There is a problem though, currently our `sort` command doesn't seem to take into account the given pattern, so it will sort all the visual selection as if no pattern was given. In order for this to work properly you need to enable Neovim `"vim.enableNeovim" = true,`. Then the sort will work properly.

We currently have an experimental support for .vimrc files that only reads mapping lines and translates them to our mappings. If you want you can try to use that, then you only need to have a line somewhere with:
```viml
cnoremap <C-k> sort /\k\S*(/ ri
```


If you really need to use the `<C-S-s>` key combination, then you can use the following workaround. Instead of creating the above remap on 'settings.json', create the following keybinding in 'keybindings.json':
```jsonc
    {
        "key": "ctrl+shift+s",
        "command": "vim.remap",
        "when": "editorTextFocus && vim.active && !inDebugRepl && vim.mode == 'CommandlineInProgress'",
        "args": {
            "after": ["s", "o", "r", "t", " ", "/", "\\", "k", "\\", "S", "*", "(", "/", " ", "r", "i"],
        }
    },
```

**Comment by TryerGit** [comment]
@berknam 

Thank you. Your first suggested method with neovim worked perfectly.

**Comment by TryerGit** [comment]
@berknam I have a question on the .vimrc support.

I enabled .vimrc support. And it asks for the path to the .vimrc. My folder structure is 

```
----projectfolder\
    ----src\
    ----include\
    ----.vscode\
           ----.vimrc
```

I fire up `code .` in `projectfolder\`

When I specified `./.vscode` it complains that there is no `.vimrc` file found there and asks me whether I should create it. Same with specification of just `.vscode` for path. I had no luck with `${workspaceFolder}/.vscode` either.

Also, what happens if there is a conflict between a mapping provided in `.vimrc` file and anything provided as settings/remappings in `settings.json` or `keybindings.json`? Which would take precedence? Also, could you confirm that if there is no conflict, the extension would take the union of `.vimrc` and the `.json`'s ?

Thank you.

**Comment by berknam** [comment]
> I enabled .vimrc support. And it asks for the path to the .vimrc. My folder structure is
> 
> ```
> ----projectfolder\
>     ----src\
>     ----include\
>     ----.vscode\
>            ----.vimrc
> ```
> 
> I fire up `code .` in `projectfolder\`
> 
> When I specified `./.vscode` it complains that there is no `.vimrc` file found there and asks me whether I should create it. Same with specification of just `.vscode` for path. I had no luck with `${workspaceFolder}/.vscode` either.

Currently it only accepts the full path to the .vimrc file, so in your case it would have to be `path-to-projectfolder\.vscode\.vimrc`.



> Also, what happens if there is a conflict between a mapping provided in `.vimrc` file and anything provided as settings/remappings in `settings.json` or `keybindings.json`? Which would take precedence? Also, could you confirm that if there is no conflict, the extension would take the union of `.vimrc` and the `.json`'s ?

The .vimrc mappings are read first so any conflicting remap made on `settings.json` will overwrite the ones on .vimrc. Any keybindings made on `keybindings.json` by the user will take priority if the `when` clause applies.

**Comment by TryerGit** [comment]
@berknam Provision of the full path gives:

vimrc file "<fullpath>/.vscode" is broken, err=Error: EISDIR: illegal operation on a directory, read

**Comment by TryerGit** [comment]
@berknam 

Setting neovim path on pure ubuntu machine worked fine.

However, on a windows machine, with WSL, after having installed neovim, `which neovim` returns `/usr/bin/nvim`. Yet entering this in ` Remote ` settings available under WSL as location `\home\user\.vscode-server\data\Machine\settings.json` gives the error:

Configuration: Invalid neovimPath. spawnSync /usr/bin/nvim ENOENT.

Is there a way to fix this on WSL ?

**Comment by berknam** [comment]
@TryerGit I believe that adding the following to your 'settings.json' file should work when using vscode in remote:
```jsonc
"remote.extensionKind": {
  "vscodevim.vim": ["workspace"]
},
```

This was addressed on #4901 but that PR was reverted with the above fix mentioned.

**Comment by TryerGit** [comment]
@berknam 

Indeed that fixes it. Prior to this fix, Ctrl-Shift-X on VSCode (opens up the extensions sidebar) showed vscodevim as being "globally enabled" only in the "local installed" tab with no option to install in remove WSL. Now, after this fix, that option was provided. Then, I had to install it in WSL, close and reopen `code .` and now it works.

Thanks!
